src/main/java/com/palantir/gradle/versions/CheckUnusedConstraintsTask.java

```diff
@@ -0,0 +1,167 @@
+import java.util.stream.Stream;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.component.ModuleComponentIdentifier;
+import org.gradle.api.artifacts.result.ResolutionResult;
+import org.gradle.api.file.RegularFile;
+import org.gradle.api.file.RegularFileProperty;
+import org.gradle.api.provider.Property;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.provider.SetProperty;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFile;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.api.tasks.options.Option;
+import org.gradle.language.base.plugins.LifecycleBasePlugin;
+
+public class CheckUnusedConstraintsTask extends DefaultTask {
+
+    private final Property<Boolean> shouldFix = getProject().getObjects().property(Boolean.class);
+    private final RegularFileProperty propsFileProperty = getProject().getObjects().fileProperty();
+    private final SetProperty<String> classpath = getProject().getObjects().setProperty(String.class);
+
+    public CheckUnusedConstraintsTask() {
+        shouldFix.set(false);
+        setGroup(LifecycleBasePlugin.VERIFICATION_GROUP);
+        setDescription("Ensures all versions in your versions.props correspond to an actual gradle dependency");
+    }
+
+    final void setPropsFile(File propsFile) {
+        this.propsFileProperty.set(propsFile);
+    }
+
+    @Input
+    public final SetProperty<String> getClasspath() {
+        return classpath;
+    }
+
+    @InputFile
+    public final Property<RegularFile> getPropsFile() {
+        return propsFileProperty;
+    }
+
+    @Option(option = "fix", description = "Whether to apply the suggested fix to versions.props")
+    public final void setShouldFix(boolean shouldFix) {
+        this.shouldFix.set(shouldFix);
+    }
+
+    final void setShouldFix(Provider<Boolean> shouldFix) {
+        this.shouldFix.set(shouldFix);
+    }
+
+    @TaskAction
+    public final void checkNoUnusedPin() {
+        Set<String> artifacts = getClasspath().get();
+        VersionsProps versionsProps = VersionsProps.loadFromFile(getPropsFile().get().getAsFile().toPath());
+
+        Set<String> exactConstraints = versionsProps.getFuzzyResolver().exactMatches();
+        Set<String> unusedConstraints = new HashSet<>(Sets.difference(exactConstraints, artifacts));
+        Set<String> unmatchedArtifacts = new HashSet<>(Sets.difference(artifacts, exactConstraints));
+
+        // assumes globs are sorted by specificity
+        for (FuzzyPatternResolver.Glob glob : versionsProps.getFuzzyResolver().globs()) {
+            Optional<String> matchingArtifact = unmatchedArtifacts.stream().filter(glob::matches).findFirst();
+            if (matchingArtifact.isPresent()) {
+                unmatchedArtifacts.remove(matchingArtifact.get());
+            } else {
+                unusedConstraints.add(glob.getRawPattern());
+            }
+        }
```

Comment:
This is not quite what I meant.
The above code 'validates' any glob that matches at least an artifact (then removing the artifact from the pool).
But a glob could match multiple artifacts, rendering other lower-priority globs unused, even if they also match some artifacts. In the code above, a second glob could come in and steal a 2nd artifact that should have matched the first glob.

Reply:
Actually this functionality is already on FuzzyResolver: 
```
unmatchedArtifacts.stream().flatMap(ua -> Stream.stream(getFuzzyResolver().patternFor(ua)))
```
gives you a list of glob.rawPattern that did match stuff, then you just need to subtract this from unusedConstraints.